Prow GitHub Client test flakes

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/test-infra/5137/pull-test-infra-bazel-canary/24/#prowgithubgo_default_test

```
--- FAIL: TestRequestRateLimit (0.04s)
	client_test.go:63: Expected to sleep for at least a second, got 999.500002ms
FAIL
```

/shrug
/kind bug
/area prow

--- a/prow/github/client.go
+++ b/prow/github/client.go
@@ -40,9 +40,24 @@ type Logger interface {
 	Debugf(s string, v ...interface{})
 }
 
+type timeClient interface {
+	Sleep(time.Duration)
+	Until(time.Time) time.Duration
+}
+
+type standardTime struct{}
+
+func (s *standardTime) Sleep(d time.Duration) {
+	time.Sleep(d)
+}
+func (s *standardTime) Until(t time.Time) time.Duration {
+	return time.Until(t)
+}
+
 type Client struct {
 	// If logger is non-nil, log all method calls with it.
 	logger Logger
+	time   timeClient
 
 	gqlc   *githubql.Client
 	client *http.Client
@@ -67,6 +82,7 @@ const (
 func NewClient(token, base string) *Client {
 	return &Client{
 		logger: logrus.WithField("client", "github"),
+		time:   &standardTime{},
 		gqlc:   githubql.NewClient(oauth2.NewClient(context.Background(), oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token}))),
 		client: &http.Client{},
 		token:  token,
@@ -81,6 +97,7 @@ func NewClient(token, base string) *Client {
 func NewDryRunClient(token, base string) *Client {
 	return &Client{
 		logger: logrus.WithField("client", "github"),
+		time:   &standardTime{},
 		gqlc:   githubql.NewClient(oauth2.NewClient(context.Background(), oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token}))),
 		client: &http.Client{},
 		token:  token,
@@ -93,6 +110,7 @@ func NewDryRunClient(token, base string) *Client {
 func NewFakeClient() *Client {
 	return &Client{
 		logger: logrus.WithField("client", "github"),
+		time:   &standardTime{},
 		fake:   true,
 		dry:    true,
 	}
@@ -108,8 +126,6 @@ func (c *Client) log(methodName string, args ...interface{}) {
 	}
 	c.logger.Debugf("%s(%s)", methodName, strings.Join(as, ", "))
 }
-
-var timeSleep = time.Sleep
 
 type request struct {
 	method      string
@@ -185,7 +201,7 @@ func (c *Client) requestRetry(method, path, accept string, body interface{}) (*h
 				// be caused by a bad API call and we'll just burn through API
 				// tokens.
 				resp.Body.Close()
-				timeSleep(backoff)
+				c.time.Sleep(backoff)
 				backoff *= 2
 			} else if resp.StatusCode == 403 {
 				if resp.Header.Get("X-RateLimit-Remaining") == "0" {
@@ -195,9 +211,9 @@ func (c *Client) requestRetry(method, path, accept string, body interface{}) (*h
 					if t, err = strconv.Atoi(resp.Header.Get("X-RateLimit-Reset")); err == nil {
 						// Sleep an extra second plus how long GitHub wants us to
 						// sleep. If it's going to take too long, then break.
-						sleepTime := time.Until(time.Unix(int64(t), 0)) + time.Second
+						sleepTime := c.time.Until(time.Unix(int64(t), 0)) + time.Second
 						if sleepTime > 0 && sleepTime < maxSleepTime {
-							timeSleep(sleepTime)
+							c.time.Sleep(sleepTime)
 						} else {
 							break
 						}
@@ -213,11 +229,11 @@ func (c *Client) requestRetry(method, path, accept string, body interface{}) (*h
 			} else {
 				// Retry 500 after a break.
 				resp.Body.Close()
-				timeSleep(backoff)
+				c.time.Sleep(backoff)
 				backoff *= 2
 			}
 		} else {
-			timeSleep(backoff)
+			c.time.Sleep(backoff)
 			backoff *= 2
 		}
 	}

--- a/prow/github/client_test.go
+++ b/prow/github/client_test.go
@@ -30,8 +30,21 @@ import (
 	"time"
 )
 
+type testTime struct {
+	now   time.Time
+	slept time.Duration
+}
+
+func (tt *testTime) Sleep(d time.Duration) {
+	tt.slept = d
+}
+func (tt *testTime) Until(t time.Time) time.Duration {
+	return t.Sub(tt.now)
+}
+
 func getClient(url string) *Client {
 	return &Client{
+		time: &standardTime{},
 		client: &http.Client{
 			Transport: &http.Transport{
 				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
@@ -42,39 +55,37 @@ func getClient(url string) *Client {
 }
 
 func TestRequestRateLimit(t *testing.T) {
-	var slept time.Duration
-	timeSleep = func(d time.Duration) { slept = d }
-	defer func() { timeSleep = time.Sleep }()
+	tc := &testTime{now: time.Now()}
 	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if slept == 0 {
+		if tc.slept == 0 {
 			w.Header().Set("X-RateLimit-Remaining", "0")
-			w.Header().Set("X-RateLimit-Reset", strconv.Itoa(int(time.Now().Add(time.Second).Unix())))
+			w.Header().Set("X-RateLimit-Reset", strconv.Itoa(int(tc.now.Add(time.Second).Unix())))
 			http.Error(w, "403 Forbidden", http.StatusForbidden)
 		}
 	}))
 	defer ts.Close()
 	c := getClient(ts.URL)
+	c.time = tc
 	resp, err := c.requestRetry(http.MethodGet, c.base, "", nil)
 	if err != nil {
 		t.Errorf("Error from request: %v", err)
 	} else if resp.StatusCode != 200 {
 		t.Errorf("Expected status code 200, got %d", resp.StatusCode)
-	} else if slept < time.Second {
-		t.Errorf("Expected to sleep for at least a second, got %v", slept)
+	} else if tc.slept < time.Second {
+		t.Errorf("Expected to sleep for at least a second, got %v", tc.slept)
 	}
 }
 
 func TestRetry404(t *testing.T) {
-	var slept int
-	timeSleep = func(d time.Duration) { slept++ }
-	defer func() { timeSleep = time.Sleep }()
+	tc := &testTime{now: time.Now()}
 	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if slept == 0 {
+		if tc.slept == 0 {
 			http.Error(w, "404 Not Found", http.StatusNotFound)
 		}
 	}))
 	defer ts.Close()
 	c := getClient(ts.URL)
+	c.time = tc
 	resp, err := c.requestRetry(http.MethodGet, c.base, "", nil)
 	if err != nil {
 		t.Errorf("Error from request: %v", err)

